### PR TITLE
Add configurations and support SSE

### PR DESCRIPTION
**Pull request by JinnyYi**
Add configurations and support SSE as describled in ref: https://github.com/aos-dev/go-service-s3/issues/51.

### CODE REVIEW

File: utils.go [hunk 1]
```diff
@@ -204,6 +225,9 @@ func (s *Service) newStorage(pairs ...typ.Pair) (st *Storage, err error) {
 	if opt.HasWorkDir {
 		st.workDir = opt.WorkDir
 	}
+	if opt.HasExceptedBucketOwner {
```
Comment (Xuanwo) [inline utils.go:228]: We don't need to set `ExceptedBucketOwner` in `Storage.New`.

File: storage.go [hunk 2]
```diff
@@ -211,11 +252,12 @@ func (s *Storage) nextPartObjectPageByPrefix(ctx context.Context, page *ObjectPa
...
-		KeyMarker:      &input.keyMarker,
-		MaxUploads:     &input.maxKeys,
-		Prefix:         &input.prefix,
-		UploadIdMarker: &input.uploadIdMarker,
+		Bucket:              &s.name,
+		KeyMarker:           &input.keyMarker,
+		MaxUploads:          &input.maxKeys,
+		Prefix:              &input.prefix,
+		UploadIdMarker:      &input.uploadIdMarker,
+		ExpectedBucketOwner: &s.exceptedBucketOwner,
```
Comment (Xuanwo) [inline storage.go:260]: If we need to set `ExpectedBucketOwner`, we should add it in `objectPageStatus` so we can construct it in list.

File: service.toml [hunk 3]
```diff
@@ -20,30 +20,92 @@ implement = ["multiparter"]
...
+description = "set this to `true` to have the S3 service client to use the region specified in the ARN, when an ARN is provided as an argument to a bucket parameter"
+
 [pairs.default_storage_pairs]
 type = "DefaultStoragePairs"
 description = "set default pairs for storager actions"
 
 [pairs.storage_class]
 type = "string"
 
+[pairs.bucket_key_enabled]
```
Comment (Xuanwo) [inline service.toml:80]: What's this pair used for?

File: service.toml [hunk 4]
```diff
@@ -20,30 +20,92 @@ implement = ["multiparter"]
...
+
+[pairs.disable_100_continue]
+type = "bool"
+description = "set this to `true` to disable the SDK adding the `Expect: 100-Continue` header to PUT requests over 2MB of content"
+
+[pairs.use_accelerate]
+type = "bool"
+description = "set this to `true` to enable S3 Accelerate feature"
+
+[pairs.disable_content_md5_validation]
```
Comment (Xuanwo) [inline service.toml:65]: Remove this pair as we always disabled it to allow users to calculate them manully.

File: service.toml [hunk 5]
```diff
@@ -20,30 +20,92 @@ implement = ["multiparter"]
...
+
+[pairs.sse_customer_algorithm]
+type = "string"
+description = "specifies the algorithm to use to when encrypting the object"
+
+[pairs.sse_customer_key]
+type = "string"
+description = "specifies the customer-provided encryption key for Amazon S3 to use to decrypt the source object"
+
+[pairs.sse_customer_key_md5]
```
Comment (Xuanwo) [inline service.toml:95]: Let's remove the `sse_customer_key_md5` and calculate for users before sending requests.

**Comment by Xuanwo** [comment]
This PR should be merged after the proposal https://github.com/aos-dev/specs/pull/38 accepted.

File: storage.go [hunk 6]
```diff
@@ -68,9 +68,8 @@ func (s *Storage) createMultipart(ctx context.Context, path string, opt pairStor
 	}
 	if opt.HasSseCustomerKey {
 		input.SSECustomerKey = &opt.SseCustomerKey
-	}
-	if opt.HasSseCustomerKeyMd5 {
-		input.SSECustomerKeyMD5 = &opt.SseCustomerKeyMd5
+		keyMD5 := ComputeMD5(opt.SseCustomerKey)
```
Comment (Xuanwo) [inline storage.go:71]: We can use `hex.EncodeToString(md5.Sum(data))`

File: service.toml [hunk 7]
```diff
@@ -23,27 +23,85 @@ required = ["location", "name"]
...
+[pairs.excepted_bucket_owner]
+type = "string"
+description = "the account ID of the excepted bucket owner"
+
+[pairs.sse_customer_algorithm]
+type = "string"
+description = "specifies the algorithm to use to when encrypting the object"
+
+[pairs.sse_customer_key]
+type = "string"
```
Comment (Xuanwo) [inline service.toml:92]: The type should be `byte_array` (upgrade go-storage to latest first)

File: storage.go [hunk 8]
```diff
@@ -52,6 +59,28 @@ func (s *Storage) createMultipart(ctx context.Context, path string, opt pairStor
...
+		input.BucketKeyEnabled = &opt.BucketKeyEnabled
+	}
+	if opt.HasExceptedBucketOwner {
+		input.ExpectedBucketOwner = &opt.ExceptedBucketOwner
+	}
+	if opt.HasSseCustomerAlgorithm {
+		input.SSECustomerAlgorithm = &opt.SseCustomerAlgorithm
+	}
+	if opt.HasSseCustomerKey {
+		customerKey := string(opt.SseCustomerKey)
```
Comment (xxchan) [inline storage.go:72]: Should this be Base-64 encoded?

File: storage.go [hunk 9]
```diff
@@ -52,6 +59,28 @@ func (s *Storage) createMultipart(ctx context.Context, path string, opt pairStor
...
+		input.SSECustomerKey = &customerKey
+		data := md5.Sum(opt.SseCustomerKey)
+		keyMD5 := hex.EncodeToString(data[:])
+		input.SSECustomerKeyMD5 = &keyMD5
```
Comment (xxchan) [inline storage.go:76]: keyMD5 should also be Base-64 encoded

Comment (xxchan) [reply storage.go:76]: We can refer to gcs's code:
```go

func setEncryptionHeaders(headers http.Header, key []byte, copySource bool) error {
	if key == nil {
		return nil
	}
	// TODO(jbd): Ask the API team to return a more user-friendly error
	// and avoid doing this check at the client level.
	if len(key) != 32 {
		return errors.New("storage: not a 32-byte AES-256 key")
	}
	var cs string
	if copySource {
		cs = "copy-source-"
	}
	headers.Set("x-goog-"+cs+"encryption-algorithm", "AES256")
	headers.Set("x-goog-"+cs+"encryption-key", base64.StdEncoding.EncodeToString(key))
	keyHash := sha256.Sum256(key)
	headers.Set("x-goog-"+cs+"encryption-key-sha256", base64.StdEncoding.EncodeToString(keyHash[:]))
	return nil
}
```

File: service.toml [hunk 10]
```diff
@@ -23,27 +23,85 @@ required = ["location", "name"]
...
+
+[pairs.sse_customer_algorithm]
+type = "string"
+description = "specifies the algorithm to use to when encrypting the object"
+
+[pairs.sse_customer_key]
+type = "byte_array"
+description = "specifies the customer-provided encryption key for Amazon S3 to use to decrypt the source object"
+
+[pairs.sse_kms_key_id]
```
Comment (Xuanwo) [inline service.toml:95]: Please update pairs' names to meet the proposal https://github.com/aos-dev/specs/blob/master/rfcs/38-service-pair-naming-style.md requirement.

`sse_kms_key_id` should be `server_side_encryption_aws_kms_key_id`.

File: storage.go [hunk 11]
```diff
@@ -349,6 +479,34 @@ func (s *Storage) write(ctx context.Context, path string, r io.Reader, size int6
...
+	if opt.HasExceptedBucketOwner {
+		input.ExpectedBucketOwner = &opt.ExceptedBucketOwner
+	}
+	if opt.HasServerSideEncryptionBucketKeyEnabled {
+		input.BucketKeyEnabled = &opt.ServerSideEncryptionBucketKeyEnabled
+	}
+	if opt.HasServerSideEncryptionCustomerAlgorithm {
+		input.SSECustomerAlgorithm = &opt.ServerSideEncryptionCustomerAlgorithm
+	}
+	if opt.HasServerSideEncryptionCustomerKey {
```
Comment (Xuanwo) [inline storage.go:491]: Maybe we can introduce a function as we dose in https://github.com/aos-dev/go-service-qingstor/pull/38/files#diff-8b7f25bac2ae04c2ce2abb628c6aa28103fd113103a62f3e9d232fa995684915R393-R402

```go
calculateEncryptionHeaders(algo string, key []byte) (algorithm, keyBase64, keyMD5Base64 *string, err error)
```

We should do the key length check in this function

File: errors.go [hunk 12]
```diff
@@ -0,0 +1,8 @@
+package s3
+
+import "errors"
+
+var (
+	// ErrInvalidEncryptionCustomerKey will be returned while encryption customer key is invalid.
+	ErrInvalidEncryptionCustomerKey = errors.New("invalid encryption customer key")
```
Comment (Xuanwo) [inline errors.go:7]: Maybe we can use `ErrServerSideEncryptionCustomerKey` here? As the input pair is called `ServerSideEncryptionCustomerKey`.

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**